File: src/agent/codex/launch.py
#!/usr/bin/env python3
# Luvus Codex launcher: carry this invocation's pane route across the daemon.
import json
import os
from pathlib import Path
import shlex
import shutil
import sys


def interactive(args):
    # Administrative commands must never seed a shared daemon with a pane route.
    values = {"-c", "--config", "-m", "--model", "-p", "--profile", "-s",
              "--sandbox", "-a", "--ask-for-approval", "-C", "--cd", "--add-dir",
              "-i", "--image", "--enable", "--disable", "--local-provider",
              "--remote-auth-token-env"}
    commands = {"agents", "exec", "e", "review", "login", "logout", "mcp",
                "plugin", "app-server", "remote-control", "completion", "update",
                "doctor", "sandbox", "debug", "apply", "queue", "archive",
                "delete", "migrate-rollouts", "unarchive", "cloud", "exec-server",
                "features", "help"}
    skip = False
    command = None
    for arg in args:
        if skip:
            skip = False
            continue
        if arg in ("-h", "--help", "-V", "--version", "--remote") or arg.startswith("--remote="):
            return False
        if arg in values:
            skip = True
        elif arg == "--":
            break
        elif not arg.startswith("-") and command is None:
            command = arg
    return command not in commands


def main():
    directory = Path(__file__).resolve().parent
    paths = [entry for entry in os.environ.get("PATH", "").split(os.pathsep)
             if Path(entry or ".").resolve() != directory]
    executable = shutil.which("codex", path=os.pathsep.join(paths))
    if not executable:
        sys.exit("Luvus: codex is not installed on the remaining PATH")
    # Nested Luvus homes can put two launchers on PATH. Each must remove itself
    # before delegating, otherwise the two wrappers would resolve each other.
    os.environ["PATH"] = os.pathsep.join(paths)
    args = sys.argv[1:]
    route = {key: os.environ.get(key, "") for key in
             ("LUVUS_ENV", "LUVUS_SOCKET_PATH", "LUVUS_PANE_ID", "LUVUS_BIN_PATH")}
    if interactive(args) and route["LUVUS_ENV"] == "1" and all(route.values()):
        home = Path(os.environ.get("CODEX_HOME") or Path.home() / ".codex")
        script = home / "luvus-agent-hook.sh"
        if script.is_file() and os.access(script, os.X_OK):
            # Session-scoped config travels with thread/start or thread/resume.
            # No daemon environment, cwd guess, or globally saved pane ID is used.
            command = shlex.join(["env", "-u", "LUVUS_SESSION",
                                  "LUVUS_CODEX_HOOK_CONTEXT=1",
                                  *[key + "=" + value for key, value in route.items()],
                                  str(script)])
            overrides = []
            for event in ("SessionStart", "UserPromptSubmit"):
                # TOML inline tables, with JSON string escaping for literal paths.
                matcher = 'matcher="startup|resume",' if event == "SessionStart" else ''
                value = '[{' + matcher + 'hooks=[{type="command",command=' + json.dumps(command, ensure_ascii=False) + ',timeout=5}]}]'
                overrides.extend(["-c", "hooks." + event + "=" + value])
            args = overrides + args
    os.execv(executable, [executable, *args])


if __name__ == "__main__":
    main()

File: src/agent/codex/launch.rs
//! Captured launch arguments must never retain a previous pane's hook route.

fn owned_hook_override(value: &str) -> bool {
    let Some((key, value)) = value.split_once('=') else {
        return false;
    };
    matches!(key, "hooks.SessionStart" | "hooks.UserPromptSubmit")
        && value.contains("LUVUS_CODEX_HOOK_CONTEXT=1")
        && value.contains("luvus-agent-hook.sh")
}

pub(in crate::agent) fn persistent_flags(args: &[String]) -> Vec<String> {
    let mut out = Vec::new();
    let mut i = 0;
    while i < args.len() {
        let arg = &args[i];
        if matches!(arg.as_str(), "-c" | "--config")
            && args.get(i + 1).is_some_and(|v| owned_hook_override(v))
        {
            i += 2;
            continue;
        }
        if arg
            .strip_prefix("--config=")
            .or_else(|| arg.strip_prefix("-c"))
            .is_some_and(|v| owned_hook_override(v.trim_start_matches('=')))
        {
            i += 1;
            continue;
        }
        out.push(arg.clone());
        i += 1;
    }
    out
}

pub(in crate::agent) fn resume_flags(args: &[String]) -> Vec<String> {
    let args = persistent_flags(args);
    let mut out = Vec::new();
    let mut i = 0;
    let mut command_seen = false;
    let mut session_pending = false;
    while i < args.len() {
        let arg = args[i].as_str();
        if arg == "--" {
            // The remainder is an initial prompt, not reusable launch options.
            break;
        }
        if matches!(
            arg,
            "-c" | "--config"
                | "-m"
                | "--model"
                | "-p"
                | "--profile"
                | "-s"
                | "--sandbox"
                | "-a"
                | "--ask-for-approval"
                | "-C"
                | "--cd"
                | "--add-dir"
                | "-i"
                | "--image"
                | "--enable"
                | "--disable"
                | "--local-provider"
                | "--remote-auth-token-env"
        ) {
            out.push(args[i].clone());
            i += 1;
            if let Some(value) = args.get(i) {
                out.push(value.clone());
                i += 1;
            }
            continue;
        }
        if !command_seen && matches!(arg, "resume" | "fork") {
            command_seen = true;
            session_pending = true;
            i += 1;
            continue;
        }
        if session_pending && matches!(arg, "--last" | "--all") {
            i += 1;
            continue;
        }
        if !arg.starts_with('-') {
            command_seen = true;
            if session_pending {
                session_pending = false;
                i += 1;
                continue;
            }
        }
        out.push(args[i].clone());
        i += 1;
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn persistent_flags_remove_only_owned_routes_including_old_snapshots() {
        // Process command tokenization may already have removed TOML quotes.
        let route = "hooks.UserPromptSubmit=[{hooks=[{command=env LUVUS_CODEX_HOOK_CONTEXT=1 /old/luvus-agent-hook.sh}]}]";
        let user = "hooks.SessionStart=[{hooks=[{command=echo user}]}]";
        for prefix in ["-c", "--config"] {
            let args = [prefix, route, "-c", user, "--model", "resume"].map(str::to_owned);
            assert_eq!(persistent_flags(&args), ["-c", user, "--model", "resume"]);
        }
        for prefix in ["-c", "-c=", "--config="] {
            assert!(persistent_flags(&[format!("{prefix}{route}")]).is_empty());
        }
        assert_eq!(
            resume_flags(&[
                "--config=model=resume".into(),
                "resume".into(),
                "--last".into(),
                "--all".into()
            ]),
            ["--config=model=resume"]
        );
    }
}
